### PR TITLE
WIP: Add file upload cookbook (refs #466)

### PR DESCRIPTION
**Pull request by marcw**
What do you think about it?

### CODE REVIEW

File: core/file-upload.md [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+
+1. Install VichUploaderBundle
+
+Install the bundle with the help of composer:
+
+```bash
+composer require vich/uploader-bundle
+```
+
+If you use Symfony Flex and Symfony4, this will create a new configuration file
```
Comment (teohhanhui) [inline core/file-upload.md]: Please remove mention of Symfony 4. Symfony Flex works with Symfony 3.4 too. Anyway, the Symfony version doesn't matter.

File: core/file-upload.md [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+     * @var null|File
+     * @Vich\UploadableField(mapping="artwork", fileNameProperty="name")
+     */
+    private $artworkFile;
+
+    /**
+     * @var null|string
+     * @ORM\Column(type="string", length=255)
+     */
+    private $name;
```
Comment (teohhanhui) [inline core/file-upload.md]: It must not be called `name`. This is not the book's name. Anyway, I think we should stick to schema.org and show an `ImageObject` example (with `contentUrl` and the corresponding `contentFile`).

File: core/file-upload.md [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+use Symfony\Component\HttpFoundation\File\File;
+use Vich\UploaderBundle\Mapping\Annotation as Vich;
+
+/**
+ * @ORM\Table(name="app_book")
+ * @ORM\Entity()
+ * @ApiResource(
+ *     collectionOperations={
+ *          "get",
+ *          "post"={"route_name"="api_book_upload"}
```
Comment (teohhanhui) [inline core/file-upload.md]: The route name is misleading as we're not uploading a book.

File: core/file-upload.md [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+use Sensio\Bundle\FrameworkExtraBundle\Configuration\Security;
+use Symfony\Bridge\Doctrine\RegistryInterface;
+use Symfony\Bundle\FrameworkBundle\Controller\AbstractController;
+use Symfony\Component\Form\FormFactoryInterface;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\Serializer\SerializerInterface;
+use Symfony\Component\Validator\Validator\ValidatorInterface;
+
+class UploadAction
```
Comment (teohhanhui) [inline core/file-upload.md]: The name is too generic.

Comment (soyuka) [reply core/file-upload.md]: I guess that's fine for docs no?

**Comment by dunglas** [comment]
Thanks for working on this! This is a very common request and it will be great to have a doc entry about file upload.

I would suggest to take a slightly different approach that will be close to what @Toflar and I are planning to add in API Platform Core at some point:

* Create a generic `/files` or `/documents` multipart endpoint to upload all documents regardless of the entity it is related with
* This endpoint will always return a `Location` header with the URL of the stored document
* Then, just set this new URL to the entity's [`contentUrl`](http://schema.org/contentUrl) property

Something like:

```
POST /files

[the binary file as a multipart HTTP request]

201 Created
Location: /<sha1-of-the-content>

POST /books

{
  "name": "The name",
  "contentUrl": "/<sha1-of-the-content>"
}
```

I think it will require just a few changes for the code, but will be more generic. WDYT @marcw?

File: core/file-upload.md [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+        $form = $this->factory->create(BookType::class, $book);
+        $form->handleRequest($request);
+        if ($form->isSubmitted() && $form->isValid()) {
+            $em = $this->doctrine->getManager();
+            $em->persist($book);
+            $em->flush();
+
+            $content = $this->serializer->serialize($book, 'jsonld', []);
+
+            return new Response($content, 201, ['Content-Type' => 'application/ld+json']);
```
[the remark this answers is not in the document]
Comment (marcw) [reply core/file-upload.md]: I'm not sure about that. I think I tried to return $book and got an error. I'll check it out.

File: core/file-upload.md (hunk 3 above)
Comment (marcw) [reply core/file-upload.md]: What do you suggest?

File: core/file-upload.md [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+        }
+
+        // This will be handled by ApiPlatform and returns a validation error.
```
Comment (dunglas) [inline core/file-upload.md]: API Platform

File: core/file-upload.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (soyuka) [reply core/file-upload.md]: Yeah I meant `$book` directly!

File: core/file-upload.md [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+     * @var null|File
+     * @Vich\UploadableField(mapping="artwork", fileNameProperty="artworkName")
+     */
+    public $artworkFile;
+
+    /**
+     * @var null|string
+     * @ORM\Column(nullable=true)
+     */
+    public $artworkName;
```
Comment (teohhanhui) [inline core/file-upload.md]: It's not the name of the artwork either. `$artworkFileName`

File: core/file-upload.md [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+recommended you [read the documentation of
+VichUploaderBundle](https://github.com/dustin10/VichUploaderBundle/blob/master/Resources/doc/index.md)
+before proceeding. It will help you get a grasp on how the bundle works, and why we use it.
+
+## Install VichUploaderBundle
+
+Install the bundle with the help of composer:
+
+```bash
+docker-compose exec api composer require vich/uploader-bundle
```
[the remark this answers is not in the document]
Comment (dunglas) [reply core/file-upload.md]: Yes, my bad

**Comment by dunglas** [comment]
Any news? It would be great to finish this one :D

File: core/file-upload.md [hunk 9]
```diff
@@ -0,0 +1,238 @@
...
+use App\Controller\MediaObjectPost;
+use ApiPlatform\Core\Annotation\ApiResource;
+use Doctrine\ORM\Mapping as ORM;
+use Symfony\Component\HttpFoundation\File\File;
+use Vich\UploaderBundle\Mapping\Annotation as Vich;
+
+/**
+ * @ORM\Table(name="app_media_object")
+ * @ORM\Entity()
+ * @ApiResource(
```
Comment (dunglas) [inline core/file-upload.md]: `iri="http://schema.org/MediaObject"`

File: core/file-upload.md [hunk 10]
```diff
@@ -0,0 +1,238 @@
...
+
+    /**
+     * @var null|string
+     * @ORM\Column(nullable=true)
+     */
+    public $filename;
+}
+```
+
+## Handle file upload
```
Comment (dunglas) [inline core/file-upload.md]: `Handling File Upload`

File: core/file-upload.md [hunk 11]
```diff
@@ -0,0 +1,238 @@
...
+
+    public function __construct(RegistryInterface $doctrine, FormFactoryInterface $factory, ValidatorInterface $validator)
+    {
+        $this->validator = $validator;
+        $this->doctrine = $doctrine;
+        $this->factory = $factory;
+    }
+
+    /**
+     * @Security("has_role('ROLE_USER')")
```
Comment (dunglas) [inline core/file-upload.md]: IIRC, `@Security` is going to be deprecated in favor of the new `@IsGranted` annotation

File: core/file-upload.md [hunk 12]
```diff
@@ -0,0 +1,238 @@
...
+use Vich\UploaderBundle\Form\Type\VichImageType;
+
+final class MediaObjectType extends AbstractType
+{
+    public function buildForm(FormBuilderInterface $builder, array $options)
+    {
+        $builder
+            // Configure each fields you want to be submitted here, like a classic form.
+            ->add('file', VichImageType::class, [
+                'label' => 'label.artwork'
```
Comment (dunglas) [inline core/file-upload.md]: `label.file`?

File: core/file-upload.md [hunk 13]
```diff
@@ -0,0 +1,238 @@
...
+    public function configureOptions(OptionsResolver $resolver)
+    {
+        $resolver->setDefaults([
+            'data_class' => MediaObject::class,
+        ]);
+    }
+}
+```
+
+## Make a request to the `/media-objects` endpoint.
```
Comment (dunglas) [inline core/file-upload.md]: Making a Request to the `/media-objects` Endpoint

File: core/file-upload.md [hunk 14]
```diff
@@ -0,0 +1,238 @@
...
+     * @var null|File
+     * @Vich\UploadableField(mapping="media_object", fileNameProperty="filename")
+     */
+    public $file;
+
+    /**
+     * @var null|string
+     * @ORM\Column(nullable=true)
+     */
+    public $filename;
```
Comment (dunglas) [inline core/file-upload.md]: Can you rename it to `contentUrl` (to match the Schema.org naming: http://schema.org/contentUrl), and add the following annotation `@ApiProperty(iri="http://schema.org/contentUrl")` with the appropriate use.

File: core/file-upload.md [hunk 15]
```diff
@@ -0,0 +1,238 @@
...
+Your `/media-objects` endpoint is now ready to receive a `POST` request with a
+file. This endpoint accepts standard `multipart/form-data` encoded data, but
+not JSON data. You will need to format your request accordingly. After posting
+your data, you will get a response looking like this:
+
+```json
+{
+    "@type": "http://schema.org/ImageObject",
+    "@id": "/media-objects/<id>",
+    "filename": "/files/<filename>",
```
Comment (dunglas) [inline core/file-upload.md]: So will be `contentUrl`

File: core/file-upload.md [hunk 16]
```diff
@@ -0,0 +1,238 @@
...
+}
+```
+
+## Link a MediaObject resource to another resource
```
Comment (dunglas) [inline core/file-upload.md]: Linking a `MediaObject` Resource to Another Resource

File: core/file-upload.md [hunk 17]
```diff
@@ -0,0 +1,238 @@
...
+
+class Book
+{
+    // ...
+    /**
+     * @var null|MediaObject
+     * @ORM\ManyToOne(targetEntity="App\Entity\MediaObject")
+     * @ORM\JoinColumn(nullable=true)
+     */
+    public $cover;
```
Comment (dunglas) [inline core/file-upload.md]: Can you name it `image` (according to Schema.org) and add the `@ApiProperty(iri="http://schema.org/Book")` annotation

File: core/file-upload.md [hunk 18]
```diff
@@ -0,0 +1,238 @@
...
+// src/Entity/Book.php
+namespace App\Entity;
+
+use App\Controller\MediaObjectPost;
+use ApiPlatform\Core\Annotation\ApiResource;
+use Doctrine\ORM\Mapping as ORM;
+use Symfony\Component\HttpFoundation\File\File;
+use Vich\UploaderBundle\Mapping\Annotation as Vich;
+
+class Book
```
Comment (dunglas) [inline core/file-upload.md]: The `@ApiResource(iri="http://schema.org/Book")` annotation is missing.
`@Entity` too

File: core/file-upload.md [hunk 19]
```diff
@@ -0,0 +1,238 @@
...
+namespace App\Entity;
+
+use App\Controller\MediaObjectPost;
+use ApiPlatform\Core\Annotation\ApiResource;
+use Doctrine\ORM\Mapping as ORM;
+use Symfony\Component\HttpFoundation\File\File;
+use Vich\UploaderBundle\Mapping\Annotation as Vich;
+
+/**
+ * @ORM\Table(name="app_media_object")
```
Comment (teohhanhui) [inline core/file-upload.md]: This is unnecessary.

**Comment by marcw** [comment]
It should be ok now. I squashed all commits for a clean merge. Please review it one last time and let me know if there is something to change.

File: core/file-upload.md [hunk 20]
```diff
@@ -0,0 +1,244 @@
...
+recommended you [read the documentation of
+VichUploaderBundle](https://github.com/dustin10/VichUploaderBundle/blob/master/Resources/doc/index.md)
+before proceeding. It will help you get a grasp on how the bundle works, and why we use it.
+
+## Install VichUploaderBundle
+
+Install the bundle with the help of composer:
+
+```bash
+docker-compose exec api php composer require vich/uploader-bundle
```
Comment (teohhanhui) [inline core/file-upload.md]: docker-compose exec ~api~ php

File: core/file-upload.md [hunk 21]
```diff
@@ -0,0 +1,244 @@
...
+to make it look like this.
+
+```yaml
+# config/packages/vich_uploader.yaml
+vich_uploader:
+    db_driver: orm
+
+    mappings:
+        media_object:
+            uri_prefix: /files
```
Comment (teohhanhui) [inline core/file-upload.md]: "Files" could be anything. Let's call it `/media` to be consistent with **Media**Object

File: core/file-upload.md [hunk 22]
```diff
@@ -0,0 +1,244 @@
...
+            upload_destination: '%kernel.project_dir%/public/files'
```
Comment (teohhanhui) [inline core/file-upload.md]: Same here. `public/media`

File: core/file-upload.md [hunk 23]
```diff
@@ -0,0 +1,244 @@
...
+In our exemple, we will create a `MediaObject` API resource. We will post files
+to this resource endpoint, and then link the newly created resource to another
+resource (in our case: Book).
+
+The `MediaObject` resource is implemented like this:
+
+```php
+<?php
+// src/Entity/MediaObject.php
+namespace App\Entity;
```
Comment (teohhanhui) [inline core/file-upload.md]: There should be a blank line before this.

File: core/file-upload.md [hunk 24]
```diff
@@ -0,0 +1,244 @@
...
+
+use App\Controller\MediaObjectPost;
```
Comment (teohhanhui) [inline core/file-upload.md]: `use` statements should be sorted.

File: core/file-upload.md [hunk 25]
```diff
@@ -0,0 +1,244 @@
...
+use Sensio\Bundle\FrameworkExtraBundle\Configuration\IsGranted;
+use Symfony\Bridge\Doctrine\RegistryInterface;
+use Symfony\Bundle\FrameworkBundle\Controller\AbstractController;
+use Symfony\Component\Form\FormFactoryInterface;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\Serializer\SerializerInterface;
+use Symfony\Component\Validator\Validator\ValidatorInterface;
+
+final class MediaObjectPost
```
Comment (teohhanhui) [inline core/file-upload.md]: `App\Action\CreateMediaObjectAction`?

File: core/file-upload.md [hunk 26]
```diff
@@ -0,0 +1,244 @@
...
+/**
+ * @ORM\Entity()
+ * @ApiResource(
+ *     iri="http://schema.org/MediaObject",
+ *     collectionOperations={
+ *          "get",
+ *          "post"={
+ *              method="POST",
+ *              path="/media-objects",
+ *              controller=MediaObjectPost::class
```
Comment (teohhanhui) [inline core/file-upload.md]: Add trailing comma.

File: core/file-upload.md [hunk 27]
```diff
@@ -0,0 +1,244 @@
...
+ *          }
```
Comment (teohhanhui) [inline core/file-upload.md]: Add trailing comma.

File: core/file-upload.md [hunk 28]
```diff
@@ -0,0 +1,244 @@
...
+ */
+class MediaObject
+{
+    // ...
+
+    /**
+     * @var null|File
+     * @Vich\UploadableField(mapping="media_object", fileNameProperty="contentUrl")
+     */
+    public $file;
```
Comment (teohhanhui) [inline core/file-upload.md]: `$contentFile`?

File: core/file-upload.md [hunk 29]
```diff
@@ -0,0 +1,244 @@
...
+ *     },
+ * )
+ * @Vich\Uploadable
+ */
+class MediaObject
+{
+    // ...
+
+    /**
+     * @var null|File
```
Comment (teohhanhui) [inline core/file-upload.md]: `File|null`

File: core/file-upload.md [hunk 30]
```diff
@@ -0,0 +1,244 @@
...
+     * @Vich\UploadableField(mapping="media_object", fileNameProperty="contentUrl")
+     */
+    public $file;
+
+    /**
+     * @var null|string
```
Comment (teohhanhui) [inline core/file-upload.md]: `string|null`

File: core/file-upload.md [hunk 31]
```diff
@@ -0,0 +1,244 @@
...
+
+/**
+ * @ORM\Entity()
+ * @ApiResource(iri="http://schema.org/Book")
+ */
+class Book
+{
+    // ...
+    /**
+     * @var null|MediaObject
```
Comment (teohhanhui) [inline core/file-upload.md]: `MediaObject|null`

File: core/file-upload.md [hunk 32]
```diff
@@ -0,0 +1,244 @@
...
+     * @ORM\ManyToOne(targetEntity="App\Entity\MediaObject")
+     * @ORM\JoinColumn(nullable=true)
+     * @ApiProperty(iri="http://schema.org/Book")
```
Comment (teohhanhui) [inline core/file-upload.md]: The `iri` is wrong here. It should be http://schema.org/image

File: core/file-upload.md [hunk 33]
```diff
@@ -0,0 +1,244 @@
...
+use App\Controller\MediaObjectPost;
+use ApiPlatform\Core\Annotation\ApiResource;
+use Doctrine\ORM\Mapping as ORM;
+use Symfony\Component\HttpFoundation\File\File;
+use Vich\UploaderBundle\Mapping\Annotation as Vich;
+
+/**
+ * @ORM\Entity()
+ * @ApiResource(
+ *     iri="http://schema.org/MediaObject",
```
Comment (teohhanhui) [inline core/file-upload.md]: CS. It should be:

```php
/**
 * @ApiResource(iri="http://schema.org/MediaObject", collectionOperations={
 *    "get",
 *    "post"={
 *        ...
 *    },
 * })
 */
```

File: core/file-upload.md [hunk 34]
```diff
@@ -0,0 +1,244 @@
...
+    {
+        $this->validator = $validator;
+        $this->doctrine = $doctrine;
+        $this->factory = $factory;
+    }
+
+    /**
+     * @IsGranted("ROLE_USER")
+     */
+    public function __invoke(Request $request)
```
Comment (teohhanhui) [inline core/file-upload.md]: Missing return type hint.

File: core/file-upload.md [hunk 35]
```diff
@@ -0,0 +1,244 @@
...
+to serve as the book cover.
+
+We first need to edit our Book resource, and add a new property called `cover`.
+
+```php
+<?php
+// src/Entity/Book.php
+namespace App\Entity;
+
+use App\Controller\MediaObjectPost;
```
Comment (teohhanhui) [inline core/file-upload.md]: `use` statements should be sorted.

File: core/file-upload.md (hunk 25 above)
Comment (marcw) [reply core/file-upload.md]: @dunglas @soyuka ?

Comment (dunglas) [reply core/file-upload.md]: I would keep it as is. Since we migrated to SF 4/Flex, we use the default directory structure (so App\Controller).

Comment (teohhanhui) [reply core/file-upload.md]: But the class name? :D

Comment (teohhanhui) [reply core/file-upload.md]: IMO `MediaObjectPost` does not make sense.

Comment (soyuka) [reply core/file-upload.md]: :+1: about the `CreateMediaObjectAction` class name.

Comment (marcw) [reply core/file-upload.md]: alright, it's changed.

**Comment by soyuka** [comment]
Good stuff, thank you very much @marcw  this will help lots of people :).

**Comment by Simperfit** [comment]
ping @marcw, it seems I cannot get this working because of the mime-type, did I missed something ?

**Comment by marcw** [comment]
@Simperfit I'm looking at it and will provide a fix for the documentation.

**Comment by marcw** [comment]
@Simperfit @dunglas See https://github.com/api-platform/docs/pull/494

**Pull request merged**